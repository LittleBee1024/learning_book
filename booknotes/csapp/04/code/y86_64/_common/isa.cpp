#include "./_common/isa.h"

#include <string>
#include <vector>
#include <string.h>

namespace
{
   const std::vector<ISA::Instr> Y86_64_INSTR = {
       {"nop", HPACK(I_NOP, F_NONE), 1, NO_ARG, 0, 0, NO_ARG, 0, 0},
       {"halt", HPACK(I_HALT, F_NONE), 1, NO_ARG, 0, 0, NO_ARG, 0, 0},
       {"rrmovq", HPACK(I_RRMOVQ, F_NONE), 2, R_ARG, 1, 1, R_ARG, 1, 0},
       /* Conditional move instructions are variants of RRMOVQ */
       {"cmovle", HPACK(I_RRMOVQ, C_LE), 2, R_ARG, 1, 1, R_ARG, 1, 0},
       {"cmovl", HPACK(I_RRMOVQ, C_L), 2, R_ARG, 1, 1, R_ARG, 1, 0},
       {"cmove", HPACK(I_RRMOVQ, C_E), 2, R_ARG, 1, 1, R_ARG, 1, 0},
       {"cmovne", HPACK(I_RRMOVQ, C_NE), 2, R_ARG, 1, 1, R_ARG, 1, 0},
       {"cmovge", HPACK(I_RRMOVQ, C_GE), 2, R_ARG, 1, 1, R_ARG, 1, 0},
       {"cmovg", HPACK(I_RRMOVQ, C_G), 2, R_ARG, 1, 1, R_ARG, 1, 0},
       /* arg1hi indicates number of bytes */
       {"irmovq", HPACK(I_IRMOVQ, F_NONE), 10, I_ARG, 2, 8, R_ARG, 1, 0},
       {"rmmovq", HPACK(I_RMMOVQ, F_NONE), 10, R_ARG, 1, 1, M_ARG, 1, 0},
       {"mrmovq", HPACK(I_MRMOVQ, F_NONE), 10, M_ARG, 1, 0, R_ARG, 1, 1},
       {"addq", HPACK(I_ALU, A_ADD), 2, R_ARG, 1, 1, R_ARG, 1, 0},
       {"subq", HPACK(I_ALU, A_SUB), 2, R_ARG, 1, 1, R_ARG, 1, 0},
       {"andq", HPACK(I_ALU, A_AND), 2, R_ARG, 1, 1, R_ARG, 1, 0},
       {"xorq", HPACK(I_ALU, A_XOR), 2, R_ARG, 1, 1, R_ARG, 1, 0},
       /* arg1hi indicates number of bytes */
       {"jmp", HPACK(I_JMP, C_YES), 9, I_ARG, 1, 8, NO_ARG, 0, 0},
       {"jle", HPACK(I_JMP, C_LE), 9, I_ARG, 1, 8, NO_ARG, 0, 0},
       {"jl", HPACK(I_JMP, C_L), 9, I_ARG, 1, 8, NO_ARG, 0, 0},
       {"je", HPACK(I_JMP, C_E), 9, I_ARG, 1, 8, NO_ARG, 0, 0},
       {"jne", HPACK(I_JMP, C_NE), 9, I_ARG, 1, 8, NO_ARG, 0, 0},
       {"jge", HPACK(I_JMP, C_GE), 9, I_ARG, 1, 8, NO_ARG, 0, 0},
       {"jg", HPACK(I_JMP, C_G), 9, I_ARG, 1, 8, NO_ARG, 0, 0},
       {"call", HPACK(I_CALL, F_NONE), 9, I_ARG, 1, 8, NO_ARG, 0, 0},
       {"ret", HPACK(I_RET, F_NONE), 1, NO_ARG, 0, 0, NO_ARG, 0, 0},
       {"pushq", HPACK(I_PUSHQ, F_NONE), 2, R_ARG, 1, 1, NO_ARG, 0, 0},
       {"popq", HPACK(I_POPQ, F_NONE), 2, R_ARG, 1, 1, NO_ARG, 0, 0},
       {"iaddq", HPACK(I_IADDQ, F_NONE), 10, I_ARG, 2, 8, R_ARG, 1, 0},
       /* this is just a hack to make the I_POP2 code have an associated name */
       {"pop2", HPACK(I_POP2, F_NONE), 0, NO_ARG, 0, 0, NO_ARG, 0, 0},
       /* For allocation instructions, arg1hi indicates number of bytes */
       {".byte", 0x00, 1, I_ARG, 0, 1, NO_ARG, 0, 0},
       {".word", 0x00, 2, I_ARG, 0, 2, NO_ARG, 0, 0},
       {".long", 0x00, 4, I_ARG, 0, 4, NO_ARG, 0, 0},
       {".quad", 0x00, 8, I_ARG, 0, 8, NO_ARG, 0, 0},
       /* empty instruction */
       {nullptr, 0, 0, NO_ARG, 0, 0, NO_ARG, 0, 0}};

   const std::vector<std::string> Y86_64_REG_NAMES = {"%rax", "%rcx", "%rdx", "%rbx", "%rsp", "%rbp", "%rsi", "%rdi",
                                                      "%r8", "%r9", "%r10", "%r11", "%r12", "%r13", "%r14"};

   const std::vector<std::string> CC_NAMES = {
       "Z=0 S=0 O=0",
       "Z=0 S=0 O=1",
       "Z=0 S=1 O=0",
       "Z=0 S=1 O=1",
       "Z=1 S=0 O=0",
       "Z=1 S=0 O=1",
       "Z=1 S=1 O=0",
       "Z=1 S=1 O=1"};
}

namespace ISA
{
   const Instr &findInstr(const char *name)
   {
      for (const auto &instr : Y86_64_INSTR)
      {
         if (strcmp(instr.name, name) == 0)
            return instr;
      }
      return Y86_64_INSTR.back(); // invalid instruction
   }

   // Return name of instruction given its encoding
   const char *decodeInstrName(int code)
   {
      for (const auto &instr : Y86_64_INSTR)
      {
         if (code == instr.code)
            return instr.name;
      }
      return "<bad>";
   }

   REG_ID findRegister(const char *name)
   {
      for (size_t i = 0; i < Y86_64_REG_NAMES.size(); i++)
      {
         if (Y86_64_REG_NAMES[i].compare(name) == 0)
            return (REG_ID)i;
      }
      return REG_ERR;
   }

   const char *getRegName(REG_ID id)
   {
      if (id >= REG_RAX && id < REG_NONE)
         return Y86_64_REG_NAMES[id].c_str();
      return "----";
   }

   const char *getCCName(cc_t c)
   {
      int ci = c;
      if (ci < 0 || ci > 7)
         return "???????????";
      else
         return CC_NAMES[c].c_str();
   }
}
