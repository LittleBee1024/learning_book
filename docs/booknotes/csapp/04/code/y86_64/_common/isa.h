#pragma once

#include <cstddef>

typedef unsigned char byte_t;
typedef long long int word_t;
typedef long long unsigned uword_t;

/******************
 * Y86_64 Registers
 ******************/
/* REG_NONE is a special one to indicate no register */
enum REG_ID : int
{
   REG_RAX,
   REG_RCX,
   REG_RDX,
   REG_RBX,
   REG_RSP,
   REG_RBP,
   REG_RSI,
   REG_RDI,
   REG_R8,
   REG_R9,
   REG_R10,
   REG_R11,
   REG_R12,
   REG_R13,
   REG_R14,
   REG_NONE = 0xF,
   REG_ERR
};

#define VALID_REG(id) (id >= 0 && id < REG_NONE)

/******************
 * Y86_64 Instructions
 ******************/
/* Different instruction types */
enum INSTR : int
{
   I_HALT,
   I_NOP,
   I_RRMOVQ,
   I_IRMOVQ,
   I_RMMOVQ,
   I_MRMOVQ,
   I_ALU,
   I_JMP,
   I_CALL,
   I_RET,
   I_PUSHQ,
   I_POPQ,
   I_IADDQ,
   I_POP2
};

/* Default function code */
enum I_FUN : int
{
   F_NONE
};

/* Different instruction argument types */
enum INSTR_ARG : int
{
   R_ARG,
   M_ARG,
   I_ARG,
   NO_ARG
};

/* Different ALU operations */
enum ALU : int
{
   A_ADD,
   A_SUB,
   A_AND,
   A_XOR,
   A_NONE
};

/* Different Jump conditions */
enum COND : int
{
   C_YES,
   C_LE,
   C_L,
   C_E,
   C_NE,
   C_GE,
   C_G
};

/* Pack itype and function into single byte */
#define HPACK(hi, lo) ((((hi)&0xF) << 4) | ((lo)&0xF))

/* Unpack byte */
#define HI4(byte) (((byte) >> 4) & 0xF)
#define LO4(byte) ((byte)&0xF)

/* Get the opcode out of one byte instruction field */
#define GET_ICODE(instr) HI4(instr)

/* Get the ALU/JMP function out of one byte instruction field */
#define GET_FUN(instr) LO4(instr)

/******************
 * Y86_64 Condition
 ******************/
typedef unsigned char cc_t;

#define GET_ZF(cc) (((cc) >> 2) & 0x1)
#define GET_SF(cc) (((cc) >> 1) & 0x1)
#define GET_OF(cc) (((cc) >> 0) & 0x1)

#define PACK_CC(z, s, o) (((z) << 2) | ((s) << 1) | ((o) << 0))
#define DEFAULT_CC PACK_CC(1, 0, 0)

namespace ISA
{
   /**
    * @brief Y86-64 instruction structure
    */
   struct Instr
   {
      const char *name;
      unsigned char code; /* Byte code for instruction+op */
      int bytes;
      INSTR_ARG arg1;
      int arg1pos;
      int arg1hi; /* 0/1 for register argument, # bytes for allocation */
      INSTR_ARG arg2;
      int arg2pos;
      int arg2hi; /* 0/1 */
   };

   /**
    * @brief Find Y86-64 instruction struction by the instruction name
    * 
    * @param name Instruction name
    * @return const Instr&, return the reference of Y86-64 instruction struction.
    *         If no instruction is found, return the empty instruction.
    */
   const Instr &findInstr(const char *name);

   /**
    * @brief Decode Y86-64 instruction name by the icode
    * 
    * @param code Instruction code number
    * @return const char*, return the instruction name. For invalid icode, return nullptr.
    */
   const char *decodeInstrName(int code);

   /**
    * @brief Find Y86-64 reigster ID by register name
    * 
    * @param name Register name
    * @return Y_REG_ID, for invalid register name, return Y_REG_ID::ERR
    */
   REG_ID findRegister(const char *name);

   /**
    * @brief Get the Y86-64 register name by register ID
    * 
    * @param id Register ID
    * @return const char*, for invalid ID, return nullptr
    */
   const char *getRegName(REG_ID id);

   /**
    * @brief Get the name of Y76-64 CC
    * 
    * @param c, the type of CC
    * @return const char* 
    */
   const char *getCCName(cc_t c);
}
