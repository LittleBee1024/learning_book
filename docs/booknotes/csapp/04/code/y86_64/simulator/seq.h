#pragma once

#include "isa.h"

/* Defined in seq.cpp which is generated from seq.hcl by hcl tool */
word_t gen_icode();
word_t gen_ifun();
word_t gen_need_regids();
word_t gen_need_valC();
word_t gen_instr_valid();
word_t gen_srcA();
word_t gen_srcB();
word_t gen_dstE();
word_t gen_dstM();
word_t gen_aluA();
word_t gen_aluB();
word_t gen_alufun();
word_t gen_set_cc();
word_t gen_mem_addr();
word_t gen_mem_data();
word_t gen_mem_read();
word_t gen_mem_write();
word_t gen_Stat();
word_t gen_new_pc();

/* Defined in sim_seq.cpp */
namespace SEQ
{

   /* SEQ variables from HCL */
   /* Intermdiate stage values that must be used by control functions */
   extern byte_t imem_icode;
   extern byte_t imem_ifun;
   extern byte_t icode;
   extern word_t ifun;
   extern word_t ra;
   extern word_t rb;
   extern word_t valc;
   extern word_t valp;
   extern bool imem_error;
   extern bool instr_valid;
   extern word_t vala;
   extern word_t valb;
   extern word_t vale;
   extern bool cond;
   extern word_t valm;
   extern bool dmem_error;
}
