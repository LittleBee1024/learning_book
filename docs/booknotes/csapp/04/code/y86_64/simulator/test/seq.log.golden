[INFO] Create SEQ Simulator
[INFO] Loaded 141 bytes code

Cycle 0. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x0
D: instr = irmovq, rA = ----, rB = %rsp, valC = 0x200, valP = 0xa
E: instr = irmovq, valC = 0x200, valA = 0x0, valB = 0x0, valP = 0xa
   srcA = ----, srcB = ----, dstE = %rsp, dstM = ----
M: instr = irmovq, Cnd = 1, valE = 0x200, valA = 0x0, valP = 0xa
   dstE = %rsp, dstM = ----
W: instr = irmovq, valP = 0x200, valE = 0x0, valM = 0xa
   dstE = %rsp, dstM = ----

Cycle 1. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0xa
D: instr = call, rA = ----, rB = ----, valC = 0x38, valP = 0x13
E: instr = call, valC = 0x38, valA = 0x0, valB = 0x200, valP = 0x13
   srcA = ----, srcB = %rsp, dstE = %rsp, dstM = ----
M: instr = call, Cnd = 1, valE = 0x1f8, valA = 0x0, valP = 0x13
   dstE = %rsp, dstM = ----
W: instr = call, valP = 0x1f8, valE = 0x0, valM = 0x13
   dstE = %rsp, dstM = ----

Cycle 2. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x38
D: instr = irmovq, rA = ----, rB = %rdi, valC = 0x18, valP = 0x42
E: instr = irmovq, valC = 0x18, valA = 0x0, valB = 0x0, valP = 0x42
   srcA = ----, srcB = ----, dstE = %rdi, dstM = ----
M: instr = irmovq, Cnd = 1, valE = 0x18, valA = 0x0, valP = 0x42
   dstE = %rdi, dstM = ----
W: instr = irmovq, valP = 0x18, valE = 0x0, valM = 0x42
   dstE = %rdi, dstM = ----

Cycle 3. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x42
D: instr = irmovq, rA = ----, rB = %rsi, valC = 0x4, valP = 0x4c
E: instr = irmovq, valC = 0x4, valA = 0x0, valB = 0x0, valP = 0x4c
   srcA = ----, srcB = ----, dstE = %rsi, dstM = ----
M: instr = irmovq, Cnd = 1, valE = 0x4, valA = 0x0, valP = 0x4c
   dstE = %rsi, dstM = ----
W: instr = irmovq, valP = 0x4, valE = 0x0, valM = 0x4c
   dstE = %rsi, dstM = ----

Cycle 4. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x4c
D: instr = call, rA = ----, rB = ----, valC = 0x56, valP = 0x55
E: instr = call, valC = 0x56, valA = 0x0, valB = 0x1f8, valP = 0x55
   srcA = ----, srcB = %rsp, dstE = %rsp, dstM = ----
M: instr = call, Cnd = 1, valE = 0x1f0, valA = 0x0, valP = 0x55
   dstE = %rsp, dstM = ----
W: instr = call, valP = 0x1f0, valE = 0x0, valM = 0x55
   dstE = %rsp, dstM = ----

Cycle 5. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x56
D: instr = irmovq, rA = ----, rB = %r8, valC = 0x8, valP = 0x60
E: instr = irmovq, valC = 0x8, valA = 0x0, valB = 0x0, valP = 0x60
   srcA = ----, srcB = ----, dstE = %r8, dstM = ----
M: instr = irmovq, Cnd = 1, valE = 0x8, valA = 0x0, valP = 0x60
   dstE = %r8, dstM = ----
W: instr = irmovq, valP = 0x8, valE = 0x0, valM = 0x60
   dstE = %r8, dstM = ----

Cycle 6. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x60
D: instr = irmovq, rA = ----, rB = %r9, valC = 0x1, valP = 0x6a
E: instr = irmovq, valC = 0x1, valA = 0x0, valB = 0x0, valP = 0x6a
   srcA = ----, srcB = ----, dstE = %r9, dstM = ----
M: instr = irmovq, Cnd = 1, valE = 0x1, valA = 0x0, valP = 0x6a
   dstE = %r9, dstM = ----
W: instr = irmovq, valP = 0x1, valE = 0x0, valM = 0x6a
   dstE = %r9, dstM = ----

Cycle 7. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x6a
D: instr = xorq, rA = %rax, rB = %rax, valC = 0x0, valP = 0x6c
E: instr = xorq, valC = 0x0, valA = 0x0, valB = 0x0, valP = 0x6c
   srcA = %rax, srcB = %rax, dstE = %rax, dstM = ----
M: instr = xorq, Cnd = 1, valE = 0x0, valA = 0x0, valP = 0x6c
   dstE = %rax, dstM = ----
W: instr = xorq, valP = 0x0, valE = 0x0, valM = 0x6c
   dstE = %rax, dstM = ----

Cycle 8. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x6c
D: instr = andq, rA = %rsi, rB = %rsi, valC = 0x0, valP = 0x6e
E: instr = andq, valC = 0x0, valA = 0x4, valB = 0x4, valP = 0x6e
   srcA = %rsi, srcB = %rsi, dstE = %rsi, dstM = ----
M: instr = andq, Cnd = 0, valE = 0x4, valA = 0x4, valP = 0x6e
   dstE = %rsi, dstM = ----
W: instr = andq, valP = 0x4, valE = 0x0, valM = 0x6e
   dstE = %rsi, dstM = ----

Cycle 9. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x6e
D: instr = jmp, rA = ----, rB = ----, valC = 0x87, valP = 0x77
E: instr = jmp, valC = 0x87, valA = 0x0, valB = 0x0, valP = 0x77
   srcA = ----, srcB = ----, dstE = ----, dstM = ----
M: instr = jmp, Cnd = 1, valE = 0x0, valA = 0x0, valP = 0x77
   dstE = ----, dstM = ----
W: instr = jmp, valP = 0x0, valE = 0x0, valM = 0x77
   dstE = ----, dstM = ----

Cycle 10. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x87
D: instr = jne, rA = ----, rB = ----, valC = 0x77, valP = 0x90
E: instr = jne, valC = 0x77, valA = 0x0, valB = 0x0, valP = 0x90
   srcA = ----, srcB = ----, dstE = ----, dstM = ----
M: instr = jne, Cnd = 1, valE = 0x0, valA = 0x0, valP = 0x90
   dstE = ----, dstM = ----
W: instr = jne, valP = 0x0, valE = 0x0, valM = 0x90
   dstE = ----, dstM = ----

Cycle 11. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x77
D: instr = mrmovq, rA = %r10, rB = %rdi, valC = 0x0, valP = 0x81
E: instr = mrmovq, valC = 0x0, valA = 0x0, valB = 0x18, valP = 0x81
   srcA = ----, srcB = %rdi, dstE = ----, dstM = %r10
M: instr = mrmovq, Cnd = 1, valE = 0x18, valA = 0x0, valP = 0x81
   dstE = ----, dstM = %r10
W: instr = mrmovq, valP = 0x18, valE = 0xd000d000d, valM = 0x81
   dstE = ----, dstM = %r10

Cycle 12. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x81
D: instr = addq, rA = %r10, rB = %rax, valC = 0x0, valP = 0x83
E: instr = addq, valC = 0x0, valA = 0xd000d000d, valB = 0x0, valP = 0x83
   srcA = %r10, srcB = %rax, dstE = %rax, dstM = ----
M: instr = addq, Cnd = 1, valE = 0xd000d000d, valA = 0xd000d000d, valP = 0x83
   dstE = %rax, dstM = ----
W: instr = addq, valP = 0xd000d000d, valE = 0x0, valM = 0x83
   dstE = %rax, dstM = ----

Cycle 13. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x83
D: instr = addq, rA = %r8, rB = %rdi, valC = 0x0, valP = 0x85
E: instr = addq, valC = 0x0, valA = 0x8, valB = 0x18, valP = 0x85
   srcA = %r8, srcB = %rdi, dstE = %rdi, dstM = ----
M: instr = addq, Cnd = 1, valE = 0x20, valA = 0x8, valP = 0x85
   dstE = %rdi, dstM = ----
W: instr = addq, valP = 0x20, valE = 0x0, valM = 0x85
   dstE = %rdi, dstM = ----

Cycle 14. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x85
D: instr = subq, rA = %r9, rB = %rsi, valC = 0x0, valP = 0x87
E: instr = subq, valC = 0x0, valA = 0x1, valB = 0x4, valP = 0x87
   srcA = %r9, srcB = %rsi, dstE = %rsi, dstM = ----
M: instr = subq, Cnd = 0, valE = 0x3, valA = 0x1, valP = 0x87
   dstE = %rsi, dstM = ----
W: instr = subq, valP = 0x3, valE = 0x0, valM = 0x87
   dstE = %rsi, dstM = ----

Cycle 15. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x87
D: instr = jne, rA = ----, rB = ----, valC = 0x77, valP = 0x90
E: instr = jne, valC = 0x77, valA = 0x0, valB = 0x0, valP = 0x90
   srcA = ----, srcB = ----, dstE = ----, dstM = ----
M: instr = jne, Cnd = 1, valE = 0x0, valA = 0x0, valP = 0x90
   dstE = ----, dstM = ----
W: instr = jne, valP = 0x0, valE = 0x0, valM = 0x90
   dstE = ----, dstM = ----

Cycle 16. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x77
D: instr = mrmovq, rA = %r10, rB = %rdi, valC = 0x0, valP = 0x81
E: instr = mrmovq, valC = 0x0, valA = 0x0, valB = 0x20, valP = 0x81
   srcA = ----, srcB = %rdi, dstE = ----, dstM = %r10
M: instr = mrmovq, Cnd = 1, valE = 0x20, valA = 0x0, valP = 0x81
   dstE = ----, dstM = %r10
W: instr = mrmovq, valP = 0x20, valE = 0xc000c000c0, valM = 0x81
   dstE = ----, dstM = %r10

Cycle 17. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x81
D: instr = addq, rA = %r10, rB = %rax, valC = 0x0, valP = 0x83
E: instr = addq, valC = 0x0, valA = 0xc000c000c0, valB = 0xd000d000d, valP = 0x83
   srcA = %r10, srcB = %rax, dstE = %rax, dstM = ----
M: instr = addq, Cnd = 1, valE = 0xcd00cd00cd, valA = 0xc000c000c0, valP = 0x83
   dstE = %rax, dstM = ----
W: instr = addq, valP = 0xcd00cd00cd, valE = 0x0, valM = 0x83
   dstE = %rax, dstM = ----

Cycle 18. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x83
D: instr = addq, rA = %r8, rB = %rdi, valC = 0x0, valP = 0x85
E: instr = addq, valC = 0x0, valA = 0x8, valB = 0x20, valP = 0x85
   srcA = %r8, srcB = %rdi, dstE = %rdi, dstM = ----
M: instr = addq, Cnd = 1, valE = 0x28, valA = 0x8, valP = 0x85
   dstE = %rdi, dstM = ----
W: instr = addq, valP = 0x28, valE = 0x0, valM = 0x85
   dstE = %rdi, dstM = ----

Cycle 19. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x85
D: instr = subq, rA = %r9, rB = %rsi, valC = 0x0, valP = 0x87
E: instr = subq, valC = 0x0, valA = 0x1, valB = 0x3, valP = 0x87
   srcA = %r9, srcB = %rsi, dstE = %rsi, dstM = ----
M: instr = subq, Cnd = 0, valE = 0x2, valA = 0x1, valP = 0x87
   dstE = %rsi, dstM = ----
W: instr = subq, valP = 0x2, valE = 0x0, valM = 0x87
   dstE = %rsi, dstM = ----

Cycle 20. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x87
D: instr = jne, rA = ----, rB = ----, valC = 0x77, valP = 0x90
E: instr = jne, valC = 0x77, valA = 0x0, valB = 0x0, valP = 0x90
   srcA = ----, srcB = ----, dstE = ----, dstM = ----
M: instr = jne, Cnd = 1, valE = 0x0, valA = 0x0, valP = 0x90
   dstE = ----, dstM = ----
W: instr = jne, valP = 0x0, valE = 0x0, valM = 0x90
   dstE = ----, dstM = ----

Cycle 21. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x77
D: instr = mrmovq, rA = %r10, rB = %rdi, valC = 0x0, valP = 0x81
E: instr = mrmovq, valC = 0x0, valA = 0x0, valB = 0x28, valP = 0x81
   srcA = ----, srcB = %rdi, dstE = ----, dstM = %r10
M: instr = mrmovq, Cnd = 1, valE = 0x28, valA = 0x0, valP = 0x81
   dstE = ----, dstM = %r10
W: instr = mrmovq, valP = 0x28, valE = 0xb000b000b00, valM = 0x81
   dstE = ----, dstM = %r10

Cycle 22. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x81
D: instr = addq, rA = %r10, rB = %rax, valC = 0x0, valP = 0x83
E: instr = addq, valC = 0x0, valA = 0xb000b000b00, valB = 0xcd00cd00cd, valP = 0x83
   srcA = %r10, srcB = %rax, dstE = %rax, dstM = ----
M: instr = addq, Cnd = 1, valE = 0xbcd0bcd0bcd, valA = 0xb000b000b00, valP = 0x83
   dstE = %rax, dstM = ----
W: instr = addq, valP = 0xbcd0bcd0bcd, valE = 0x0, valM = 0x83
   dstE = %rax, dstM = ----

Cycle 23. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x83
D: instr = addq, rA = %r8, rB = %rdi, valC = 0x0, valP = 0x85
E: instr = addq, valC = 0x0, valA = 0x8, valB = 0x28, valP = 0x85
   srcA = %r8, srcB = %rdi, dstE = %rdi, dstM = ----
M: instr = addq, Cnd = 1, valE = 0x30, valA = 0x8, valP = 0x85
   dstE = %rdi, dstM = ----
W: instr = addq, valP = 0x30, valE = 0x0, valM = 0x85
   dstE = %rdi, dstM = ----

Cycle 24. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x85
D: instr = subq, rA = %r9, rB = %rsi, valC = 0x0, valP = 0x87
E: instr = subq, valC = 0x0, valA = 0x1, valB = 0x2, valP = 0x87
   srcA = %r9, srcB = %rsi, dstE = %rsi, dstM = ----
M: instr = subq, Cnd = 0, valE = 0x1, valA = 0x1, valP = 0x87
   dstE = %rsi, dstM = ----
W: instr = subq, valP = 0x1, valE = 0x0, valM = 0x87
   dstE = %rsi, dstM = ----

Cycle 25. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x87
D: instr = jne, rA = ----, rB = ----, valC = 0x77, valP = 0x90
E: instr = jne, valC = 0x77, valA = 0x0, valB = 0x0, valP = 0x90
   srcA = ----, srcB = ----, dstE = ----, dstM = ----
M: instr = jne, Cnd = 1, valE = 0x0, valA = 0x0, valP = 0x90
   dstE = ----, dstM = ----
W: instr = jne, valP = 0x0, valE = 0x0, valM = 0x90
   dstE = ----, dstM = ----

Cycle 26. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x77
D: instr = mrmovq, rA = %r10, rB = %rdi, valC = 0x0, valP = 0x81
E: instr = mrmovq, valC = 0x0, valA = 0x0, valB = 0x30, valP = 0x81
   srcA = ----, srcB = %rdi, dstE = ----, dstM = %r10
M: instr = mrmovq, Cnd = 1, valE = 0x30, valA = 0x0, valP = 0x81
   dstE = ----, dstM = %r10
W: instr = mrmovq, valP = 0x30, valE = 0xa000a000a000, valM = 0x81
   dstE = ----, dstM = %r10

Cycle 27. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x81
D: instr = addq, rA = %r10, rB = %rax, valC = 0x0, valP = 0x83
E: instr = addq, valC = 0x0, valA = 0xa000a000a000, valB = 0xbcd0bcd0bcd, valP = 0x83
   srcA = %r10, srcB = %rax, dstE = %rax, dstM = ----
M: instr = addq, Cnd = 1, valE = 0xabcdabcdabcd, valA = 0xa000a000a000, valP = 0x83
   dstE = %rax, dstM = ----
W: instr = addq, valP = 0xabcdabcdabcd, valE = 0x0, valM = 0x83
   dstE = %rax, dstM = ----

Cycle 28. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x83
D: instr = addq, rA = %r8, rB = %rdi, valC = 0x0, valP = 0x85
E: instr = addq, valC = 0x0, valA = 0x8, valB = 0x30, valP = 0x85
   srcA = %r8, srcB = %rdi, dstE = %rdi, dstM = ----
M: instr = addq, Cnd = 1, valE = 0x38, valA = 0x8, valP = 0x85
   dstE = %rdi, dstM = ----
W: instr = addq, valP = 0x38, valE = 0x0, valM = 0x85
   dstE = %rdi, dstM = ----

Cycle 29. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x85
D: instr = subq, rA = %r9, rB = %rsi, valC = 0x0, valP = 0x87
E: instr = subq, valC = 0x0, valA = 0x1, valB = 0x1, valP = 0x87
   srcA = %r9, srcB = %rsi, dstE = %rsi, dstM = ----
M: instr = subq, Cnd = 0, valE = 0x0, valA = 0x1, valP = 0x87
   dstE = %rsi, dstM = ----
W: instr = subq, valP = 0x0, valE = 0x0, valM = 0x87
   dstE = %rsi, dstM = ----

Cycle 30. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x87
D: instr = jne, rA = ----, rB = ----, valC = 0x77, valP = 0x90
E: instr = jne, valC = 0x77, valA = 0x0, valB = 0x0, valP = 0x90
   srcA = ----, srcB = ----, dstE = ----, dstM = ----
M: instr = jne, Cnd = 0, valE = 0x0, valA = 0x0, valP = 0x90
   dstE = ----, dstM = ----
W: instr = jne, valP = 0x0, valE = 0x0, valM = 0x90
   dstE = ----, dstM = ----

Cycle 31. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x90
D: instr = ret, rA = ----, rB = ----, valC = 0x0, valP = 0x91
E: instr = ret, valC = 0x0, valA = 0x1f0, valB = 0x1f0, valP = 0x91
   srcA = %rsp, srcB = %rsp, dstE = %rsp, dstM = ----
M: instr = ret, Cnd = 1, valE = 0x1f8, valA = 0x1f0, valP = 0x91
   dstE = %rsp, dstM = ----
W: instr = ret, valP = 0x1f8, valE = 0x55, valM = 0x91
   dstE = %rsp, dstM = ----

Cycle 32. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x55
D: instr = ret, rA = ----, rB = ----, valC = 0x0, valP = 0x56
E: instr = ret, valC = 0x0, valA = 0x1f8, valB = 0x1f8, valP = 0x56
   srcA = %rsp, srcB = %rsp, dstE = %rsp, dstM = ----
M: instr = ret, Cnd = 1, valE = 0x200, valA = 0x1f8, valP = 0x56
   dstE = %rsp, dstM = ----
W: instr = ret, valP = 0x200, valE = 0x13, valM = 0x56
   dstE = %rsp, dstM = ----

Cycle 33. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x13
D: instr = halt, rA = ----, rB = ----, valC = 0x0, valP = 0x14
E: instr = halt, valC = 0x0, valA = 0x1f8, valB = 0x1f8, valP = 0x14
   srcA = %rsp, srcB = %rsp, dstE = %rsp, dstM = ----
M: instr = halt, Cnd = 1, valE = 0x200, valA = 0x1f8, valP = 0x14
   dstE = %rsp, dstM = ----
W: instr = halt, valP = 0x200, valE = 0x13, valM = 0x14
   dstE = %rsp, dstM = ----

After 34 cycles, the state becomes HALT

Changes to registers:
%rax:	0x0000000000000000	0x0000abcdabcdabcd
%rsp:	0x0000000000000000	0x0000000000000200
%rdi:	0x0000000000000000	0x0000000000000038
%r8:	0x0000000000000000	0x0000000000000008
%r9:	0x0000000000000000	0x0000000000000001
%r10:	0x0000000000000000	0x0000a000a000a000
Changes to memory:
0x01f0:	0x0000000000000000	0x0000000000000055
0x01f8:	0x0000000000000000	0x0000000000000013
